src/sentry/replays/usecases/reader.py

```diff
@@ -0,0 +1,256 @@
+    """Return a recording segment storage metadata instance."""
+    # TODO: Filestore is privileged until the direct storage is released to all projects.  Once
+    # direct-storage is the default driver we need to invert this.  90 days after deployment we
+    # need to remove filestore querying.
+    segment = fetch_filestore_segment_meta(project_id, replay_id, segment_id)
+    if segment:
+        return segment
+
+    return fetch_direct_storage_segment_meta(project_id, replay_id, segment_id)
+
+
+def fetch_filestore_segments_meta(
+    project_id: int,
+    replay_id: str,
+    offset: int,
+    limit: int,
+) -> List[RecordingSegmentStorageMeta]:
+    """Return filestore metadata derived from our Postgres table."""
+    segments: List[ReplayRecordingSegment] = (
+        ReplayRecordingSegment.objects.filter(project_id=project_id, replay_id=replay_id)
+        .order_by("segment_id")
+        .all()[offset : limit + offset]
+    )
+    return [
+        RecordingSegmentStorageMeta(
+            project_id=project_id,
+            replay_id=replay_id,
+            segment_id=segment.segment_id,
+            retention_days=0,
+            date_added=segment.date_added,
+            file_id=segment.file_id,
+        )
+        for segment in segments
+    ]
+
+
+def fetch_filestore_segment_meta(
+    project_id: int,
+    replay_id: str,
+    segment_id: int,
+) -> RecordingSegmentStorageMeta | None:
+    """Return filestore metadata derived from our Postgres table."""
+    segment = ReplayRecordingSegment.objects.filter(
+        project_id=project_id,
+        replay_id=replay_id,
+        segment_id=segment_id,
+    ).first()
+    if segment is None:
+        return None
+
+    return RecordingSegmentStorageMeta(
+        project_id=project_id,
+        replay_id=replay_id,
+        segment_id=segment.segment_id,
+        retention_days=0,
+        date_added=segment.date_added,
+        file_id=segment.file_id,
+    )
+
+
+def fetch_direct_storage_segments_meta(
+    project_id: int,
+    replay_id: str,
+    offset: int,
+    limit: int,
+) -> List[RecordingSegmentStorageMeta]:
+    """Return direct-storage metadata derived from our Clickhouse table."""
+    if not has_archived_segment(project_id, replay_id):
+        return _fetch_segments_from_snuba(project_id, replay_id, offset, limit)
+    return []
+
+
+def fetch_direct_storage_segment_meta(
+    project_id: int,
+    replay_id: str,
+    segment_id: int,
+) -> RecordingSegmentStorageMeta | None:
+    """Return direct-storage metadata derived from our Clickhouse table."""
+    if has_archived_segment(project_id, replay_id):
+        return None
+
+    results = _fetch_segments_from_snuba(
+        project_id=project_id,
+        replay_id=replay_id,
+        offset=0,
+        limit=1,
+        segment_id=segment_id,
+    )
+    if len(results) == 0:
+        return None
+    else:
+        return results[0]
+
+
+def has_archived_segment(project_id: int, replay_id: str) -> bool:
```

Comment:
i wonder if it's worth attempting to combine this with the below query. should be possible. (we should also update our deletion logic to use the timestamp of the first segment so that time wise the deletion row is grouped with the others)

Reply:
The segment lookup query isn't grouped so this would not be possible (at least not without a sub-query which is restricted).  With project_id and replay_id conditions the query takes 250ms to run.  With a timestamp filter we can lower that to 100ms.  With an index on replay_id perhaps that could be lowered even further.  Whatever we do will benefit both queries.

I'm seeing sub 500ms response times with our current filestore implementation so without further optimizations we can expect to see a regression in performance.